translated/tech/20190810-go-monitor-pattern.md

```diff
@@ -0,0 +1,219 @@
+# Go: 监控模式
+
+![](https://raw.githubusercontent.com/studygolang/gctt-images/master/go-monitor-pattern/1.png)
+
+Go 能实现[监控模式](https://en.wikipedia.org/wiki/Monitor_(synchronization))，归功于 `sync` 包和 `sync.Cond` 结构体。监控模式允许 goroutine 在进入睡眠模式前等待一个定特定条件，而不会阻塞执行或消耗资源。
+
+## 条件变量
+我们举个例子，来看看这个模式可以带来的好处。我将使用 [Bryan Mills 的演示文稿](https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view)中提供的示例：
+
+```go
+
+type Item = int
+
+type Queue struct {
+	items     []Item
+	*sync.Cond
+}
+
+func NewQueue() *Queue {
+	q := new(Queue)
+	q.Cond = sync.NewCond(&sync.Mutex{})
+	return q
+}
+
+func (q *Queue) Put(item Item) {
+	q.L.Lock()
+	defer q.L.Unlock()
+	q.items = append(q.items, item)
+	q.Signal()
+}
+
+func (q *Queue) GetMany(n int) []Item {
+	q.L.Lock()
+	defer q.L.Unlock()
+	for len(q.items) < n {
+		q.Wait()
+	}
+	items := q.items[:n:n]
+	q.items = q.items[n:]
+	return items
+}
+
+func main() {
+	q := NewQueue()
+
+	var wg sync.WaitGroup
+	for n := 10; n > 0; n-- {
+		wg.Add(1)
+		go func(n int) {
+			items := q.GetMany(n)
+			fmt.Printf("%2d: %2d\n", n, items)
+			wg.Done()
+		}(n)
+	}
+
+	for i := 0; i < 100; i++ {
+		q.Put(i)
+	}
+
+	wg.Wait()
+}
+```
+
+`Queue` 是一个非常简单的结体构，由一个切片和 `sync.Cond` 结构组成。然后，我们做两件事：
+
+- 启动 10 个 goroutines，并将尝试一行消费 X 个元素。如果这些元素不够数目，那么 goroutine 将进去睡眠状态并等待被唤醒
+- 主 goroutine 将用 100 个元素填满队列。对于添加的元素，它将唤醒满足条件的正在睡眠的一个 goroutine。
```

Comment:
The main goroutine will fill the queue up with one hundred items.
主 goroutine 将用 100 个元素填满队列
fill：填满--填入

For each item added, it will wake up one of the goroutines that is waiting for items.
对于添加的元素，它将唤醒满足条件的正在睡眠的一个 goroutine。
每添加一个 item， 即唤醒一个正在等待 item 的 goroutine